File: python/techai_webutils/src/techai_webutils/core/interfaces/workflow.py
"""Workflow interfaces for business process orchestration.

Defines the abstract contracts for workflow implementations (sync and async).
Workflows coordinate multi-step operations across services, pipelines, and caches.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class Workflow[In, Out](ABC):
    """Abstract workflow for synchronous business process orchestration.

    Workflows coordinate multiple operations across services, pipelines, repositories,
    and caches. Unlike pipelines (stateless transforms), workflows orchestrate stateful
    business processes with conditional logic, error recovery, and cross-cutting concerns.

    Examples: Cache-aside pattern (check cache → compute → store), order fulfillment
    (verify → charge → ship → notify).

    Phase 1: Single-execution orchestration with error handling.
    Phase 2+: Long-running workflows, saga patterns, state machines.
    """

    @abstractmethod
    def execute(self, input_data: In) -> Out:
        """Orchestrate a multi-step business process.

        Implementations may be stateful and typically depend on services, repos, and caches.
        """
        ...


class AsyncWorkflow[In, Out](ABC):
    """Abstract workflow for asynchronous business process orchestration.

    Async variant of Workflow for orchestrating async operations. This is the primary
    variant for Python services which are typically async.

    Examples: Search workflow (check cache → run pipeline → cache result), ingestion
    workflow (parse → validate → transform → index), notification dispatch.
    """

    @abstractmethod
    async def execute(self, input_data: In) -> Out:
        """Orchestrate a multi-step business process asynchronously.

        Implementations may be stateful and typically depend on async services, repos, and caches.
        """
        ...

File: go/core/interfaces/workflow.go
package interfaces

import "context"

// Workflow is the generic business process orchestration interface for multi-step coordination.
// In = input type, Out = output type.
//
// Workflows coordinate multiple operations across services, pipelines, repositories, and caches.
// Unlike pipelines (stateless transforms), workflows orchestrate stateful business processes
// with conditional logic, error recovery, and cross-cutting concerns.
//
// Examples: a lookup workflow (check cache → run pipeline → cache result), ingestion workflow
// (parse → validate → transform → index), order fulfillment (verify → charge → ship → notify).
type Workflow[In any, Out any] interface {
	// Execute orchestrates a multi-step business process.
	// Context is provided for cancellation and tracing.
	// Implementations may be stateful and typically depend on services, repos, and caches.
	Execute(ctx context.Context, input In) (Out, error)
}
